---
layout: default
title: Overview and Installation
nav_order: 1
description: "NetBID 2.0 overview and installation"
permalink: /
---

      
# NetBID 2.0: Data-driven Network-based Bayesian Inference of Drivers
{: .fs-9 }

Documentation and Guided Analyses
{: .fs-6 .fw-300 }

[Get started now](#getting-started){: .btn .btn-primary .fs-5 .mb-4 .mb-md-0 .mr-2 } [View it on GitHub](https://github.com/jyyulab/NetBID-dev){: .btn .fs-5 }

---

## Overview

NetBID is a data-driven system biology pipeline, using data-driven network-based Bayesian inference approach to find drivers from transcriptomics, proteomics and phosphoproteomics data, where the drivers can be either transcription facotrs (**TF**) or signaling factors (**SIG**).

NetBID 2.0 is an upgraded version of [NetBID 1.0](https://github.com/jyyulab/NetBID/releases/tag/1.0.0) that has been published in [Nature]((https://www.nature.com/articles/s41586-018-0177-0)) in 2018. NetBID 2.0 inherites all the main functions from NetBID 1.0, and provides many more functions and pipelines to perform advanced end-to-end analyses.

**More data processing functions:** 

- Expression matrix pre-processing and quality assessment
- SJARACNe-based network construction
- Activity calculation of drivers and gene sets
- Discovery of differential expressed genes and differential activated drivers
- Generation of the master table for drivers
- For more data processing functions, please check NetBID2 PDF manual

**More visualization functions:**

- Unsupervised learning of samples, comparison between the predicted labels vs. the observed labels
- Display drivers with significance profiles and target genes
- Display selected drivers with more details and its sub-network structure
- For more visualization functions, please check NetBID2 PDF manual

**More supporting functions:**

- Gene/transcript ID conversion
- Gene function enrichment analysis & visualization
- Data and pipeline management
- For more supporting functions, please check NetBID2 PDF manual


---

## Installation

### Dependencies

R, version >= 3.6.0

[Pre-request R packages](docs/pre_request)


### Install a released R package

Download a released version from [https://github.com/jyyulab/NetBID/releases](https://github.com/jyyulab/NetBID/releases) and run:

```R
devtools::install_local('NetBID2_2.0.1.tar.gz')
```


### Install from github master branch
Clone the repository and install locally:

```R
devtools::install(pkg='.',dependencies=TRUE) ## Install the package with dependencies.
devtools::install_deps(pkg = ".", dependencies = TRUE) ## Install package dependencies if needed.
```

Install without cloning the repository

```R
library(devtools)
library(BiocManager)
# set repos, for R version 3.6.0, Bioconductor version 3.9
local({
  r <- getOption("repos")
  r["CRAN"] <- "https://cran.rstudio.com/"
  r["BioCsoft"] <- "https://bioconductor.org/packages/3.9/bioc"
  r["BioCann"] <- "https://bioconductor.org/packages/3.9/data/annotation"
  r["BioCexp"] <- "https://bioconductor.org/packages/3.9/data/experiment"
  options(repos = r)
})

devtools::install_github("jyyulab/NetBID-dev",ref='master',dependencies='Depends') 
```

---

## Design manual

The manual of all the NetBID2 functions is linked here [NetBID_manual.pdf](https://github.com/jyyulab/NetBID-dev/blob/master/NetBID_manual.pdf). 
Every function has its own demo scripts to demonstrate its functionality.

All these 70 functions can be grouped into their corresponding function modules,

![function group](function_group.png)

**Here are 4 functions we think are ESSENTIAL for the NetBID2 pipeline management:**

- `NetBID.network.dir.create()` helps users to create an organized working directory for the network construction step in NetBID2 analysis. It creates a hierarchcial working directory and returns a list containing this directory information. 

- `NetBID.analysis.dir.create()`  helps users to create an organized working directory for the driver estimation step in NetBID2 analysis. It also creates a hierarchcial working directory and returns a list contains this directory information.

- `NetBID.saveRData()` helps users to save complicated list object generated by certain steps of NetBID2â€™s pipeline. This function makes an easier data/pipeline management and reference.

- `NetBID.loadRData()` pairs with `NetBID.saveRData()`, it reloads previous saved Rdata, so users don't need to re-run everything.

**Here are 2 list objects we think are ESSENTIAL for the NetBID2 pipeline analysis:**

- `network.par` is a variable in the **network construction part**. It is created by `NetBID.network.dir.create()` with network construction directory information wrapped inside. This list object is used to store all the important results from the network construction pipeline. It can be saved as RData and reloaded back using `NetBID.saveRData()` and  `NetBID.loadRData()`.
- `analysis.par` is a variable in the **driver analysis part**. It is created by `NetBID.analysis.dir.create()` with driver estimation directory information wrapped inside. This list object is used to store all the important results from the driver estimation pipeline. It also can be saved as RData and reloaded back using `NetBID.saveRData()` and  `NetBID.loadRData()`.

The workflow of each NetBID2 pipeline is summarized in the figure below,

![pipeline part](pipeline_part.png)

**We highly suggest new users to follow the pipeline for driver analysis and visualization.** Details of the NetBID2 pipeline is shown and explained using a demo dataset in the following tutorial.
Of course, most functions in NetBID2 are still flexible to use and perform specific needs for users.


## Tutorial
 
We choose the demo dataset from GEO database: [GSE116028](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE116028). 

This microarray dataset contains 13 adult medulloblastoma (MB) samples. 
Three phenotype subgroups of adult MB have been identified from distinguishabled expression profiles, clinical features, pathological features and prognosis.
These subgroups together with their sample numbers are, 3 SHH, 4 WNT, and 6 Group4.
Group4 tumors in adult have significantly worse progression-free and overall survival, compared to other molecular subtypes of tumor.
Here, the goal is to **find potential drivers in Group4 compared to other subtypes using NetBID2**. This may relate to specific clinical feature of Group4 MB subtype.
 
**Though the dataset in the tutorial is microarray, NetBID2 is also capable to analyze RNA-Seq data. We will show in the tutorial below as well.**  
 
The tutorial contains three main parts, they can be followed by order or used independently:

1. [Network construction](docs/network_construction)

2. [Driver estimation (master table generation)](docs/driver_estimation)

3. [Advanced analysis (result visualization)](docs/advanced_analysis)

---

## About the project

For the detailed description of NetBID algorithm, please check our lab page [View Yu Lab@St. Jude](https://stjuderesearch.org/site/lab/yu).
Email: xinran.dong@stjude.org or xinran.dong@foxmail.com

### License

Distributed by an MIT license.
